### PR TITLE
[ROCm] Enable gemm fusion autotuner.

### PR DESCRIPTION
**Pull request by zoranjovanovic-ns**

### CODE REVIEW

File: xla/service/gpu/autotuning/gemm_fusion_autotuner_test.cc [hunk 1]
```diff
@@ -306,13 +323,26 @@ class GemmFusionAutotunerTest : public StatelessAutotunerTest {
     return debug_options;
   }
 
+  stream_executor::GpuComputeCapability CudaAmpereOrRocm() {
```
Comment (dimitar-asenov) [inline xla/service/gpu/autotuning/gemm_fusion_autotuner_test.cc:326]: This function can be somewhat confusing. I suggest you simplify the logic and directly inline it in the single place where it is used:

```
if (IsRocm()) {
return GetRocmComputeCapability();
} else {
...
}
```
I've another comment suggesting you introduce `IsRocm()`, and it's probably also good to have `GetRocmComputeCapability()`, similarly to how we have a function for cuda.

File: xla/service/gpu/autotuning/gemm_fusion_autotuner_test.cc [hunk 2]
```diff
@@ -1260,6 +1316,10 @@ TEST_F(GemmFusionAutotunerTest,
 }
 
 TEST_F(GemmFusionAutotunerTest, RewritesGemmFusionToCustomKernelFusion) {
+  if (std::holds_alternative<se::RocmComputeCapability>(
```
Comment (dimitar-asenov) [inline xla/service/gpu/autotuning/gemm_fusion_autotuner_test.cc:1319]: This pattern appears quite often. Please make a peer to `GpuComputeComp` called `IsRocm` that does the check and use it everywhere instead.

File: xla/service/gpu/autotuning/gemm_fusion_autotuner_test.cc (hunk 1 above)
Comment (dimitar-asenov) [reply xla/service/gpu/autotuning/gemm_fusion_autotuner_test.cc:326]: Once you introduce those two new function, please make a pass over the entire file and use them where appropriate.

File: xla/service/gpu/autotuning/BUILD [hunk 3]
```diff
@@ -28,7 +32,9 @@ package_group(
 
 cc_library(
     name = "gemm_fusion_autotuner",
-    srcs = ["gemm_fusion_autotuner.cc"],
+    srcs = ["gemm_fusion_autotuner.cc"] +
+            if_cuda_is_configured(["gemm_fusion_autotuner_cuda.cc"]) +
+            if_rocm_is_configured(["gemm_fusion_autotuner_rocm.cc"]),
     hdrs = ["gemm_fusion_autotuner.h"],
     tags = [
         "cuda-only",
```
Comment (dimitar-asenov) [inline xla/service/gpu/autotuning/BUILD:40]: You should remove this tag.

File: xla/service/gpu/autotuning/gemm_fusion_autotuner.h [hunk 4]
```diff
@@ -198,6 +197,26 @@ class GemmFusionAutotunerImpl {
...
+  int block_k = 0;
+};
+
+absl::StatusOr<TileSizeLimit> GetLimits(const HloDotInstruction& dot);
+bool IsFusionKind(const HloInstruction& hlo, absl::string_view kind);
+
+// Minimum tile size.
+constexpr int kMinTileSize = 16;
+
+using BackendConfig = GemmFusionAutotunerImpl::BackendConfig;
```
Comment (dimitar-asenov) [inline xla/service/gpu/autotuning/gemm_fusion_autotuner.h:212]: Please do not put `using` declarations in header files.

File: xla/service/gpu/autotuning/gemm_fusion_autotuner.h [hunk 5]
```diff
@@ -198,6 +197,26 @@ class GemmFusionAutotunerImpl {
   std::vector<TritonGemmConfig> triton_configs_;
 };
 
+struct TileSizeLimit {
```
Comment (dimitar-asenov) [inline xla/service/gpu/autotuning/gemm_fusion_autotuner.h:200]: All of the declarations below are currently directly in the `xla::gpu` namespace, but shouldn't be. The reason is that their name are too generic for top-level symbols.

Would moving these as members of `GemmFusionAutotuner` and/or `GemmFusionAutotunerImpl` also work?

File: xla/service/gpu/autotuning/gemm_fusion_autotuner_rocm.cc [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+  TF_ASSIGN_OR_RETURN(std::vector<TritonGemmConfig> triton_configs,
+                      GenerateTritonConfigs(*dot));
+  for (TritonGemmConfig& config : triton_configs) {
+    configs.push_back(std::move(config));
+  }
+  return configs;
+}
+
+absl::StatusOr<std::vector<TritonGemmConfig>>
+GemmFusionAutotunerImpl::GenerateTritonConfigs(const HloDotInstruction& dot) {
```
Comment (dimitar-asenov) [inline xla/service/gpu/autotuning/gemm_fusion_autotuner_rocm.cc:215]: This method is too long, please break it down into multiple methods.

File: xla/service/gpu/autotuning/gemm_fusion_autotuner_cuda.cc [hunk 7]
```diff
@@ -0,0 +1,383 @@
+/* Copyright 2024 The OpenXLA Authors.
```
Comment (dimitar-asenov) [inline xla/service/gpu/autotuning/gemm_fusion_autotuner_cuda.cc:1]: I looked at the diff between this file and the `_rocm` version. At the moment it is very small and I think we'll be much better off if we keep this a single file with a few places where the code branches. Also in general, I suspect that's desirable for the long-term.

Do you have a strong reason why splitting this code and duplicating a lot of it is desirable? If not, please revert this change.

File: xla/service/gpu/autotuning/gemm_fusion_autotuner.h (hunk 5 above)
Comment (dimitar-asenov) [reply xla/service/gpu/autotuning/gemm_fusion_autotuner.h:200]: Also look at my comment on the new `_cuda` file. Perhaps we can simply revert the change here as well, as suggested in that comment.

**Review by dimitar-asenov — CHANGES_REQUESTED** [review]

File: xla/service/gpu/autotuning/gemm_fusion_autotuner_cuda.cc (hunk 7 above)
Comment (zoranjovanovic-ns) [reply xla/service/gpu/autotuning/gemm_fusion_autotuner_cuda.cc:1]: The reason for creating two files (_cuda and _rocm versions) is that there are parts of the code that cannot be compiled as a part of cuda or rocm build (e.g. missing cuda header files in rocm build).
Keeping it in single file would require introduction of #if GOOGLE_CUDA (or similar) guards.
My understanding is that introduction of these new #ifdef is not acceptable,

File: xla/service/gpu/autotuning/gemm_fusion_autotuner_rocm.cc (hunk 6 above)
Comment (zoranjovanovic-ns) [reply xla/service/gpu/autotuning/gemm_fusion_autotuner_rocm.cc:215]: Just copied existing corresponding Cuda method. If it necessary, I will break it to several smaller methods.

File: xla/service/gpu/autotuning/gemm_fusion_autotuner_cuda.cc (hunk 7 above)
Comment (dimitar-asenov) [reply xla/service/gpu/autotuning/gemm_fusion_autotuner_cuda.cc:1]: You are correct, we should not introduce new `#ifdef` macros. However, my comment still mostly stands. We need a hybrid solution:

- Leave all common code in the current file.
- Isolate only parts that require Rocm/Cud headers into the new `_rocm`, `_cuda` files. If I look at the diff between the two, I believe that's just:
  - `const int64_t BLAS_GEMM_DEFAULT`
  - `GetCuDnnPlanCount` and more generally the CuDNN bit of code.
- Everything else could stay in one file. E.g. for `GetDefaultTritonConfigs()` the code can just add the common configs unconditionally, then check for cuda and add the remaining bits.

@beckerhe For having two different implementations of a common header (`rocm` and `cuda`), is the current solution OK (different `srcs` based on `if_rocm/cuda_configured`)? Is it preferable to have two entirely separate targets that are then conditionally added to `deps`, e.g. like we do [here](https://github.com/openxla/xla/blob/main/xla/mlir/tools/mlir_interpreter/dialects/BUILD#L73-L110)?

Comment (beckerhe) [reply xla/service/gpu/autotuning/gemm_fusion_autotuner_cuda.cc:1]: > Is it preferable to have two entirely separate targets that are then conditionally added to deps, e.g. like we do [here](https://github.com/openxla/xla/blob/main/xla/mlir/tools/mlir_interpreter/dialects/BUILD#L73-L110)?

Yes, that would be preferrable for a couple of reasons:

1. Both targets can have different dependencies and different build flags if needed
2. Dependency cleaners like `build_cleaner` only work if these are separate targets

**Comment by zoranjovanovic-ns** [comment]
@dimitar-asenov Pushed new version

File: xla/service/gpu/autotuning/gemm_fusion_autotuner_rocm.cc (hunk 6 above)
Comment (dimitar-asenov) [reply xla/service/gpu/autotuning/gemm_fusion_autotuner_rocm.cc]: Right, the original method is also too long. If you end up modifying the original (if there's just one left after the proposals) or if you fork it, please break it down. It's good to do minor clean ups as we touch old code.

File: xla/service/gpu/autotuning/BUILD [hunk 8]
```diff
@@ -31,10 +31,11 @@ package_group(
 )
 
 cc_library(
-    name = "gemm_fusion_autotuner",
-    srcs = ["gemm_fusion_autotuner.cc"] +
-            if_cuda_is_configured(["gemm_fusion_autotuner_cuda.cc"]) +
-            if_rocm_is_configured(["gemm_fusion_autotuner_rocm.cc"]),
+    name = "gemm_fusion_autotuner_cuda",
+    srcs = [
+        "gemm_fusion_autotuner.cc",
```
[the remark this answers is not in the document]
Comment (beckerhe) [reply xla/service/gpu/autotuning/BUILD]: Thank you. This looks good to me. (Only looked a the build rule things)

[the remark this answers is not in the document]
Comment (zoranjovanovic-ns) [reply xla/service/gpu/autotuning/BUILD]: @dimitar-asenov Modified build rules.

File: xla/service/gpu/autotuning/gemm_fusion_autotuner.cc [hunk 9]
```diff
@@ -1267,33 +1241,45 @@ GemmFusionAutotunerImpl::GetExhaustiveTritonConfigs() const {
...
-      Config(16, 64, 256, 8, 1, 4),   Config(256, 256, 128, 1, 3, 8)};
-  if (GetComputeCapability().IsAtLeastHopper()) {
-    absl::c_copy(
-        std::vector<Config>{
-            Config(16, 32, 32, 8, 1, 2),
-            Config(16, 64, 128, 8, 1, 4),
-            Config(16, 64, 128, 16, 3, 4),
-        },
-        std::back_inserter(configs));
+  if(isRocm()) {
```
Comment (dimitar-asenov) [inline xla/service/gpu/autotuning/gemm_fusion_autotuner.cc:1244]: As far as I can see all the configs here apply to both Rocm and Cuda. Please remove the IsRocm codition and just add these undconditionally. Then below, you can have an `isCuda` and add only what's relevant there.

That makes it a bit more obvious what is common and what not.

File: xla/service/gpu/autotuning/gemm_fusion_autotuner.h [hunk 10]
```diff
@@ -183,11 +185,21 @@ class GemmFusionAutotunerImpl {
...
-  se::CudaComputeCapability GetComputeCapability() const {
-    return std::get<se::CudaComputeCapability>(
-        config_.GetGpuComputeCapability());
+  se::GpuComputeCapability GetComputeCapability() const {
+    return config_.GetGpuComputeCapability();
+  }
+
+  bool isRocm() const {
+    return std::holds_alternative<se::RocmComputeCapability>(
+          GetComputeCapability()) ? true : false;
```
Comment (dimitar-asenov) [inline xla/service/gpu/autotuning/gemm_fusion_autotuner.h:194]: Please remove `? true : false`, the result of `holds_alternative` is already what you want.

File: xla/service/gpu/autotuning/gemm_fusion_autotuner_rocm.cc [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "xla/service/gpu/autotuning/gemm_fusion_autotuner.h"
+#include "xla/stream_executor/rocm/rocm_blas.h"
+
+#include <algorithm>
```
Comment (dimitar-asenov) [inline xla/service/gpu/autotuning/gemm_fusion_autotuner_rocm.cc:19]: Please remove all unused headers (and corresponding deps in the `BUILD` file).

File: xla/service/gpu/autotuning/gemm_fusion_autotuner_test.cc [hunk 12]
```diff
@@ -207,6 +211,26 @@ class StatelessAutotunerTest : public HloTestBase {
...
+        .rocm_compute_capability();
+  }
+
+  const stream_executor::GpuComputeCapability& GpuComputeComp() {
+    return backend().default_stream_executor()
+                    ->GetDeviceDescription().gpu_compute_capability();
+  }
+
+  bool isRocm() {
+    if (std::holds_alternative<se::RocmComputeCapability>(
```
Comment (dimitar-asenov) [inline xla/service/gpu/autotuning/gemm_fusion_autotuner_test.cc:227]: The entire body of this function should just be:
```
return std::holds_alternative<se::RocmComputeCapability>(
          GpuComputeComp());
```

**Review by dimitar-asenov — CHANGES_REQUESTED** [review]

File: xla/service/gpu/autotuning/gemm_fusion_autotuner.cc (hunk 9 above)
Comment (zoranjovanovic-ns) [reply xla/service/gpu/autotuning/gemm_fusion_autotuner.cc:1244]: I would like to separate parts of the code for adding configs for rocm and cuda.
I do expect adding/removing new configs for rocm and would like to avoid changing cuda related code when adding/removing configs for rocm.

Comment (dimitar-asenov) [reply xla/service/gpu/autotuning/gemm_fusion_autotuner.cc:1244]: In that case, please implement this function separately in the `_cuda` and `_rocm` cc files, since the logic for the two cases will be completely disjoint.

**Review by dimitar-asenov — CHANGES_REQUESTED** [review]

**Review by dimitar-asenov — APPROVED** [review]

**Comment by dimitar-asenov** [comment]
Thank you for patiently addressing all of our comments!
